### DISCUSSION TITLE
Best Practice Large Queries

### DISCUSSION THREAD

**Discussion by Marvin-TMG** [godror/godror#170] [Q&A] [answered]
Hi all,

I am new to go, so my apologies if my questions are misguided. I need to process large queries with an Oracle database, which can not all be easily combined. I am looking for some suggestions on how to best implement this.

What I have currently:
- API end points created with Fiber
- Service which combines all the necessary function calls
- Series of query functions using sqlx and godror driver

This is my end point, pretty simple for now:
```
	app.Get("/get_data", func(c *fiber.Ctx) error {
		return get_data(c, db)
	})
```

This is my service, I am using channels and go routines here for concurrency (not sure if this is correct):
```
func get_data(c *fiber.Ctx, db *sqlx.DB) error {
	start := time.Now()
       // collection of different data structures we need to query independently
	var d DataStruct

	// Make channels
	ch1 := make(chan []data_struct_1)
        ...
	ch7 := make(chan []data_struct_2)

	// Setup goroutines to fetch data
	input_1 := 118370
	input_2 := "some_string"

	go query_db_1(db, ch1, input_1)
        ...
	go query_db_2(db, ch7, input_2)

	// Accumulate results
	d.Results1 = <-ch1
         ...
	d.Results2 = <-ch7

	fmt.Printf("Total query time %s\n", time.Since(start))

	return c.JSON(d)
}
```

Each query function is something like the following, I am using godror.PrefetchCount and FetchArraySize as additional arguments to increase the through put.
```
func query_db_1(db *sqlx.DB, c chan []data_struct_1, input_1 int) {
	start := time.Now()

	data := []data_struct_1{}
	err := db.Select(&data, `SELECT t.ID, t.NODEDEF_ID, t.SENT, t.VALUE 
							FROM table_name t WHERE t.CATEGORY_ID = :1`, input_1,
		godror.PrefetchCount(5001), godror.FetchArraySize(5000))

	if err != nil {
		fmt.Println(err)
		close(c)
	}
	fmt.Printf("Ints Query %s\n", time.Since(start))

	c <- data
	close(c)
}
```

What I am noticing is that if I increase the amounts of calls, the query time simply increases. So this indicates to me that there is not much work down in parallel. I wonder if this can be improved somehow or if you have any general suggestions how to structure such queries?

One final issue I have, is that I need to convert a custom object stored in the database, I am not sure how to accomplish this with Godror, I currently do the following, but it seems convoluted:
```
err = db.Select(&data, `SELECT t.ID, t.NODEDEF_ID, t.SENT, t.VALUERAW, t.BOUNDERS
								FROM table_name t WHERE t.CATEGORY_ID = :1`, input_1,
		godror.PrefetchCount(5001), godror.FetchArraySize(5000))

	if err != nil {
		fmt.Println(err)
		close(c)
	}

	for i, blob := range data {
		var datatype []int64
		obj := blob.BOUNDERS.(*godror.Object)
		val, _ := obj.Collection().AsSlice(datatype)
		data[i].BOUNDERS = val
	}
```

Thank you for your feedback!

**Comment by sudarshan12s** [comment]
on the parallel queries, you have some stats on Go pool (db.Stats) , OCI Session pool stats(GetPoolStats)  printed to see if goroutines are in wait states. 
Does it improve after tuning pool parameters like poolMaxSessions, poolMinSessions, poolIncrement ,?

**Reply by Marvin-TMG** [reply]
Thanks for the info. I was able to use the db.Stats, but I could figure out how to obtain GetPoolStats.

If I understand correctly, the general process as described above is a correct implementation for running parallel queries? the sql or sqlx module automatically takes care of distributing these requests over different connections? Is it generally recommended to run queries in a go routine or rather use normal functions?

I overwrote the following parameters:
db.SetMaxIdleConns(20)
db.SetMaxOpenConns(20)

This allowed me to see the amount of connections increase beyond 2 (it went up to 6, when manually pressing F5 quickly in the browser). 
```
{
        "MaxOpenConnections": 20,
        "OpenConnections": 6,
        "InUse": 0,
        "Idle": 6,
        "WaitCount": 0,
        "WaitDuration": 0,
        "MaxIdleClosed": 0,
        "MaxIdleTimeClosed": 0,
        "MaxLifetimeClosed": 0
}
```

**Comment by tgulacsi** [comment, marked as answer]
What does sqlx.DB.Select do? Collect all the rows into a slice of given struct?
Then this is more or less okay. Using channels here is not required. 

```
   import "golang.org/x/sync/errgroup"
...
  var grp errgroup.Group
  grp.Go(func() error { return db.Select(&db.Result1, `SELECT t.ID, t.NODEDEF_ID, t.SENT, t.VALUE 
							FROM table_name t WHERE t.CATEGORY_ID = :1`, input_1,
		godror.PrefetchCount(5001), godror.FetchArraySize(5000)) })
...
  return grp.Wait()
```

is more straightforward, faster and more parallel.

**Reply by Marvin-TMG** [reply]
Thanks for the suggestion, that is very helpfull I will try it asap! Yes, sqlx.Select is a convenience wrapper for the standard database, providing direct slicing to structures (saves a bit of manual mapping work).

One final question: what is the preferred way to convert convert Oracle Objects to for example a list of integers (this is what our custom object represents). I currently do the following, but it is a bit inconvenient:
```
err = db.Select(&data, `SELECT t.ID, t.NODEDEF_ID, t.SENT, t.VALUERAW, t.BOUNDERS FROM table_name t WHERE t.CATEGORY_ID = :1`, input_1)

// convert BOUNDERS from Object to []int
	for i, blob := range data {
		var datatype []int64
		obj := blob.BOUNDERS.(*godror.Object)
		val, _ := obj.Collection().AsSlice(datatype)
		data[i].BOUNDERS = val
	}
```
In the cx_Oracle Python library, one can define an output conversion function in Python, which converts the objects on the fly. Is something like this available in godror as well?

**Reply by tgulacsi** [reply]
That code is OK.
You can create a type with proper Scan method to have the Select/Scan fill it, but I think it's more readable to have it explicitly.

**Reply by Marvin-TMG** [reply]
Thanks, this is working fine now!

I also indeed see the connections increasing up to 20 and then the wait counter is increasing (which would lead to longer response times). So all seems to be working as expected for now.
